### PR TITLE
ENH: format sample sheets

### PR DESCRIPTION
**Pull request by wasade**
Using the Illumina Experiment Manager, I was able to load sheets for:

- target gene on miseq
- target gene on hiseq
- shotgun on miseq
- shotgun on hiseq

### CODE REVIEW

File: knimin/lib/format.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        The email of an additional contact.
+
+    Sample Sheet Note
+    -----------------
+    If the instrument type is a MiSeq, any lane information per-sample will be
+    disregarded. If instrument type is a HiSeq, each sample must include a
+    "lane" key.
+
+    If the run type is Target Gene, then sample details are disregarded
+    with the exception of determining the the lanes.
```
Comment (josenavas) [inline knimin/lib/format.py]: the the  -> the

File: knimin/lib/format.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        else:
+            sample_details = [sample_detail_fmt % {'run_name': run_name}]
+    else:
+        sample_details = [sample_detail_fmt % samp
+                          for samp in sample_information]
+
+    base_sheet = _format_general(run_name, labadmin_id, assay, fwd_cycles,
+                                 rev_cycles, pi_name, pi_email,
+                                 contact_0_name, contact_0_email,
+                                 contact_1_name, contact_1_email,
```
Comment (josenavas) [inline knimin/lib/format.py]: I think this should be c1name, c1email, etc...

File: knimin/lib/format.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        sample_header = DATA_TARGET_GENE_STRUCTURE
+        sample_detail = DATA_TARGET_GENE_SAMPLE_STRUCTURE
+    elif run_type == 'Shotgun':
+        sample_header = DATA_SHOTGUN_STRUCTURE
+        sample_detail = DATA_SHOTGUN_SAMPLE_STRUCTURE
+    else:
+        raise ValueError("%s is not a known run type" %
+                         run_type)
+
+    # if its a miseq, there isn't lane information
```
Comment (qiyunzhu) [inline knimin/lib/format.py]: This configuration is so specific. In the future it will inevitably need modification.

File: knimin/lib/format.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    sample_detail_fmt = sample_prefix + sample_detail + sample_suffix
+
+    if run_type == 'Target Gene':
+        if instrument_type == 'hiseq':
+            lanes = sorted({samp['lane'] for samp in sample_information})
+            sample_details = []
+            for idx, lane in enumerate(lanes):
+                # make a unique run-name on the assumption this is required
+                detail = {'lane': lane, 'run_name': run_name + str(idx)}
+                sample_details.append(sample_detail_fmt % detail)
```
Comment (qiyunzhu) [inline knimin/lib/format.py]: I didn't know that one can do `%` in a variable :)

File: knimin/lib/format.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    optional = {
+        'contact_1_name': contact_1_name,
+        'contact_1_email': contact_1_email,
+        'contact_2_name': contact_2_name,
+        'contact_2_email': contact_2_email
+    }
+
+    for k, v in fmt.items():
```
Comment (qiyunzhu) [inline knimin/lib/format.py]: Not important, I just wonder if this long function can be simplified by things like `locals` and `exec` with a for loop.

File: knimin/lib/format.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    elif run_type == 'Shotgun':
+        sample_header = DATA_SHOTGUN_STRUCTURE
+        sample_detail = DATA_SHOTGUN_SAMPLE_STRUCTURE
+    else:
+        raise ValueError("%s is not a known run type" %
+                         run_type)
+
+    # if its a miseq, there isn't lane information
+    if instrument_type == 'miseq':
+        header_prefix = ''
```
Comment (qiyunzhu) [inline knimin/lib/format.py]: Trivial. I guess that one can do `header_prefix  = sample_prefix  = ''`

File: knimin/lib/format.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    c1email = contact_1_email if contact_1_email is not None else ''
+    c2name = contact_2_name if contact_2_name is not None else ''
+    c2email = contact_2_email if contact_2_email is not None else ''
+
+    validate_contact(pi_name, pi_email)
+    validate_contact(contact_0_name, contact_0_email)
+    validate_contact(contact_1_name, contact_1_email)
+    validate_contact(contact_2_name, contact_2_email)
+
+    if fwd_cycles <= 0 or not isinstance(fwd_cycles, int):
```
Comment (qiyunzhu) [inline knimin/lib/format.py]: Is `fwd_cycles` a value that the user can type in the interface? I wonder if it's necessary to validate it in the backend.

Comment (wasade) [inline knimin/lib/format.py]: The backend will drive this code

File: knimin/lib/format.py (hunk 3 above)
Comment (wasade) [inline knimin/lib/format.py]: Possibly. We only actually care about a few of the fields in here though, same with the instrument

File: knimin/lib/format.py (hunk 6 above)
Comment (wasade) [inline knimin/lib/format.py]: meh

File: knimin/lib/format.py (hunk 4 above)
Comment (wasade) [inline knimin/lib/format.py]: :)

File: knimin/lib/format.py (hunk 5 above)
Comment (wasade) [inline knimin/lib/format.py]: There are likely introspective tricks to be had but it may also make it more difficult to understand.

**Pull request merged**